### PR TITLE
fix(core): dynamic page a11y

### PR DESCRIPTION
**Pull request by genereg1**
## Related Issue(s)

closes https://github.com/SAP/fundamental-ngx/issues/5576

## Description

- fixes dynamic Page accessibility issues. 

##### During Implementation

1. Visual Testing:

-   [x] visual misalignments/updates
-   [x] check Light/Dark/HCB/HCW themes
-   [x] RTL/LTR - proper rendering and labeling
-   [x] responsiveness(resize)
-   [x] Content Density (Cozy/Compact/(Condensed))
-   [x] States - hover/disabled/focused/active/on click/selected/selected hover/press state
-   [x] Interaction/Animation - open/close, expand/collapse, add/remove, check/uncheck
-   [x] Mouse vs. Keyboard support
-   [x] Text Truncation

2. API and functional correctness

-   [x] check for console logs (warnings, errors)
-   [n/a] API boundary values
-   [x] different combinations of components - free style
-   [n/a] change the API values during testing

3. Documentation and Example validations

-   [x] missing API documentation or it is not understandable
-   [x] poor examples
-   [x] Stackblitz works for all examples

4. Accessibility testing
5. Browser Testing - Edge, Safari, Chrome, Firefox

##### PR Quality

-   [x] the commit message(s) follows the guideline:
        https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
-   [x] tests for the changes that have been done
-   [x] all items on the PR Review Checklist are addressed :
        https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
-   [x] Run npm run build-pack-library and test in external application
-   [n/a] update `README.md`
-   [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)

### CODE REVIEW

File: libs/core/src/lib/breadcrumb/breadcrumb.interface.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { InjectionToken } from '@angular/core';
+
+export interface BreadcrumbInterface {
```
Comment (dimamarksman) [inline libs/core/src/lib/breadcrumb/breadcrumb.interface.ts:3]: Let's BreadcrumbComponent implement this

File: libs/core/src/lib/breadcrumb/breadcrumb-item.directive.ts [hunk 2]
```diff
@@ -45,4 +76,35 @@ export class BreadcrumbItemDirective implements AfterContentInit {
...
+    /** @hidden */
+    ngAfterViewInit(): void {
+        this._arrowNavigation = this._breadcrumbComponent.arrowNavigation;
+    }
+
+    @HostListener('keydown', ['$event'])
+    handleKeydown(event: KeyboardEvent): void {
+        if (this._arrowNavigation) {
+            const nextElement: HTMLElement = this.elementRef.nativeElement.nextElementSibling.children[0];
+            const isNextElementLink = nextElement && nextElement.hasAttribute('href');
```
Comment (dimamarksman) [inline libs/core/src/lib/breadcrumb/breadcrumb-item.directive.ts:89]: As for me nextElementSibling.children looks not flexible enough.

Maybe it's better to listen for this event from breadcrumb component and move focus to this item if needed?
The item itself just has to implement .focus() method.
Moreover breadcrumb can rely on [FocusKeyManger service](https://material.angular.io/cdk/a11y/overview#focuskeymanager)

File: libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html [hunk 3]
```diff
@@ -13,10 +13,10 @@
         <ng-container *ngTemplateOutlet="startColumn"></ng-container>
     </div>
     <!-- Left Separator -->
-    <div *ngIf="_leftColumnSeparator" class="fd-flexible-column-layout__separator">
+    <div *ngIf="_leftColumnSeparator" class="fd-flexible-column-layout__separator" aria-label="separator">
```
Comment (Betrozov) [inline libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html:16]: please add internationalization for `aria-label`

File: libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html [hunk 4]
```diff
@@ -13,10 +13,10 @@
...
         <button
             class="fd-button fd-button--compact fd-button--transparent fd-flexible-column-layout__button"
-            aria-label="Expand"
+            [title]="_leftColumnSeparator === 'left' ? 'Expand' : 'Collapse'"
```
Comment (Betrozov) [inline libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html:19]: same for `title`

File: libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html [hunk 5]
```diff
@@ -35,10 +35,10 @@
         <ng-container *ngTemplateOutlet="midColumn"></ng-container>
     </div>
     <!-- Right Separator -->
-    <div *ngIf="_rightColumnSeparator" class="fd-flexible-column-layout__separator">
+    <div *ngIf="_rightColumnSeparator" class="fd-flexible-column-layout__separator" aria-label="separator">
```
Comment (Betrozov) [inline libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html:38]: same

File: libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html [hunk 6]
```diff
@@ -35,10 +35,10 @@
...
         <button
             class="fd-button fd-button--compact fd-button--transparent fd-flexible-column-layout__button"
-            aria-label="Expand"
+            [title]="_rightColumnSeparator === 'left' ? 'Expand' : 'Collapse'"
```
Comment (Betrozov) [inline libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html:41]: same

**Review by Betrozov — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/breadcrumb/breadcrumb.component.ts [hunk 7]
```diff
@@ -92,26 +88,50 @@ export class BreadcrumbComponent implements AfterContentInit, OnInit, OnDestroy
...
     private _subscriptions = new Subscription();
 
+    /** @hidden */
+    private _keyManager: FocusKeyManager<BreadcrumbItemDirective>;
+
     constructor(
         public elementRef: ElementRef,
         @Optional() private _rtlService: RtlService,
         @Optional() private _contentDensityService: ContentDensityService,
+        @Optional() private _dynamicPageService: DynamicPageService,
```
Comment (dimamarksman) [inline libs/core/src/lib/breadcrumb/breadcrumb.component.ts]: Breadcrumb component shouldn't know anything about DynamicPage

**Review by dimamarksman — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html [hunk 8]
```diff
@@ -13,10 +13,16 @@
...
+    <div
+        *ngIf="_leftColumnSeparator"
+        class="fd-flexible-column-layout__separator"
+        i18n-aria-label="coreFlexibleColumnLayout.Left.Separator|Separator aria-label attribute"
+        aria-label="separator"
+    >
         <button
             class="fd-button fd-button--compact fd-button--transparent fd-flexible-column-layout__button"
-            aria-label="Expand"
+            i18n-title="coreFlexibleColumnLayount.Left.Separator.Toogle|Left title attribute"
```
Comment (Betrozov) [inline libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html:24]: please use `@Input` for internationalization
please pay attention to this issue
https://github.com/SAP/fundamental-ngx/issues/6436

File: libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html [hunk 9]
```diff
@@ -35,10 +41,16 @@
...
+    <div
+        *ngIf="_rightColumnSeparator"
+        class="fd-flexible-column-layout__separator"
+        i18n-aria-label="coreFlexibleColumnLayout.Right.Separator|Separator aria-label attribute"
+        aria-label="separator"
+    >
         <button
             class="fd-button fd-button--compact fd-button--transparent fd-flexible-column-layout__button"
-            aria-label="Expand"
+            i18n-title="coreFlexibleColumnLayount.Right.Separator.Toogle|Right title attribute"
```
Comment (Betrozov) [inline libs/core/src/lib/flexible-column-layout/flexible-column-layout.component.html:52]: same

**Review by Betrozov — CHANGES_REQUESTED** [review]

**Comment by genereg1** [comment]
> Focusing order is broken when open a dynamic page https://deploy-preview-7345--fundamental-ngx.netlify.app/#/core/dynamic-page#basic
> 
>  Screen.Recording.2021-12-23.at.16.06.13.mov

@platon-rov 
It's intended behavior. For now, you can navigate through Breadcrumbs by arrow keys according to the doc where it says that tab navigation should be reduced. Therefore Tab key on Breadcrumbs in Dynamic Page just switches to another UI element. I added `[arrowNavigation]` property to Breadcrumbs it handles how you prefer to navigate through Breadcrumbs by Tab or Arrow keys

**Review by Betrozov — APPROVED** [review]

**Review by dmitry-stepanenko — APPROVED** [review]

**Review by platon-rov — APPROVED** [review]

**Comment by Veerakarthick** [comment]
Hello Colleague,

Retest has been taken results are below:

Issue1: Fixed
Issue2: Fixed
Issue3:
Observed Behavior:
On navigating to the tabs, an additional information ‘tab’ is read at first on entering the tabstrip section. Also, the tutor message provided to change the tab is not precise. i.e. it reads to change the selection use ctrl+page down but the tabs cannot be switched using the provided shortcut key. Therefore, the information read is not precise.
The speech output is:
�Tab� ��
Tab 1 �Tab� �selected� ��
1 of 3
To switch pages, press Control+PageDown.
Tab 2 �Tab�
2 of 3
To switch pages, press Control+PageDown.

Issue4: Fixed
Issue5: Fixed
Issue6: Fixed
Issue7: Fixed
Issue8: Fixed
Issue9: Fixed
Issue10: Fixed

Retest environment:
URL:https://deploy-preview-7345--fundamental-ngx.netlify.app/#/core/dynamic-page[](https://user-images.githubusercontent.com/37998745/160749392-4812ced4-5a8c-4a69-85c4-d351606ba2e2.png)
Screen reader: Jaws 2021.2111.13 ILM
Browser: Version 99.0.4844.84 (Official Build) (64-bit)

Regards,
Karthikeyan.c

image

**Comment by Karthickc1994** [comment]
Hello Colleague,

Retest has been taken results are below:

Issue1: Fixed
Issue2: Fixed
Issue3:
Observed Behavior:
On navigating to the tabs, an additional information ‘tab’ is read at first on entering the tabstrip section. Also, the tutor message provided to change the tab is not precise. i.e. it reads to change the selection use ctrl+page down but the tabs cannot be switched using the provided shortcut key. Therefore, the information read is not precise.
The speech output is:
�Tab� ��
Tab 1 �Tab� �selected� ��
1 of 3
To switch pages, press Control+PageDown.
Tab 2 �Tab�
2 of 3
To switch pages, press Control+PageDown.

screen shot:
image

Issue4: Fixed
Issue5: Fixed
Issue6: Fixed
Issue7: Fixed
Issue8: Fixed
Issue9: Fixed
Issue10: Fixed

Retest environment:
URL:https://fundamental-ngx-gh--pr7345-fix-dynamic-page-a11-45ql36g8.web.app/fundamental-ngx#/core/dynamic-page
Browser: chrome Version 102.0.5005.115 (Official Build) (64-bit)
screen reader: Jaws2022.2204.20ILM

Regards,
Karthikeyan.c

**Comment by Karthickc1994** [comment]
Hello Colleague,

The reported issue in CASE 3  is belongs to combination is a known JAWS bug in Chrome issue.
Enter into a tablist JAWS narrates: To switch pages, press Control+PageDown https://github.com/FreedomScientific/VFO-standards-support/issues/337

so this PR will be closed with technology exception.
 
Regards,
Karthikeyan.c

**Comment by genereg1** [comment]
> Hello Colleague,
> 
> The reported issue in CASE 3 is belongs to combination is a known JAWS bug in Chrome issue. Enter into a tablist JAWS narrates: To switch pages, press Control+PageDown [FreedomScientific/VFO-standards-support#337](https://github.com/FreedomScientific/VFO-standards-support/issues/337)
> 
> so this PR will be closed with technology exception.
> 
> Regards, Karthikeyan.c

Hello Kartrick,
Does it mean this PR is verified and we can merge it?

**Pull request merged**